### PR TITLE
Implement aggressive_removal metric in terms of set operations

### PR DESCRIPTION
**Pull request by mdittmer**
Depends on https://github.com/foam-framework/foam2/pull/317 #56 for passing tests.

### CODE REVIEW

File: lib/confluence/aggressive_removal.es6.js [hunk 1]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
-                      numAggressiveRemoval,
-                      date,
-                    }));
-              });
-            });
-          }));
-        }
-        return Promise.all(computeForBrowsers);
+        // Select from first current browser.
+        return firstCurrent.interfaces.select(
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: I think the reason that this is handled specially should be documented.

File: lib/confluence/aggressive_removal.es6.js [hunk 2]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+            // Intersect with other current browsers to get APIs shipped by
+            // all (non-"browser" vendor) current browsers.
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: Suggestion: Intersect with interfaces of other browsers in "current" to get a list of interfaces that exist in all of them.

File: lib/confluence/aggressive_removal.es6.js [hunk 3]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+    },
+    {
+      name: 'computeFromBrowsers',
+      documentation: `Compute aggressive removal value for "browser", given
+          arrays of "current" browsers from other vendors and "previous"
+          browsers (from "browser" vendor) that are older than the grace
+          period.
+
+          Let browserAPIs = <"browser" interfaces>
+          Let curAPIs = INTERSECT(<"current" interfaces>)
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: suggestion: interfaces for each item in "current"

File: lib/confluence/aggressive_removal.es6.js [hunk 4]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+      name: 'computeForBrowser',
+      documentation: 'Compute aggressive removal value for a single browser.',
+      code: function(browsers, date, browser) {
+        let browserKeys = browsers.map((browser) => browser.browserKey);
+        return Promise.all([
+          // Current browsers from vendor other than "browser".
+          this.browserDAO.where(this.mlang.AND(
+              this.mlang.NEQ(this.Browser.BROWSER_NAME, browser.browserName),
+              this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys)))
+                  .orderBy(this.mlang.DESC(this.Browser.RELEASE_DATE))
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: How does this produce just the latest version for each browser key? Is it missing a group by somewhere?

browserDAO has a function getLatestBrowserFromEachVendorAtDate. Is it possible to use that here?

File: lib/confluence/aggressive_removal.es6.js [hunk 5]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+            browser => this.computeForBrowser(browsers, date, browser)));
+      },
+    },
+    {
+      name: 'computeForBrowser',
+      documentation: 'Compute aggressive removal value for a single browser.',
+      code: function(browsers, date, browser) {
+        let browserKeys = browsers.map((browser) => browser.browserKey);
+        return Promise.all([
+          // Current browsers from vendor other than "browser".
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: As discussed offline, I think we need a terminology refresh here. 'browser' should refer to a product (Chrome, Edge, etc.) and 'vendor' should refer to the organization responsible for one or more browsers (Google, Microsoft, etc.). Some vendors have had multiple browsers in the past.

File: lib/confluence/aggressive_removal.es6.js [hunk 6]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+          this.browserDAO.where(this.mlang.AND(
+              this.mlang.NEQ(this.Browser.BROWSER_NAME, browser.browserName),
+              this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys)))
+                  .orderBy(this.mlang.DESC(this.Browser.RELEASE_DATE))
+                  .select(),
+          // Browsers released before the grace period "browser" vendor.
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: suggestion: ...by browser's vendor

File: lib/confluence/aggressive_removal.es6.js [hunk 7]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+      name: 'computeFromBrowsers',
+      documentation: `Compute aggressive removal value for "browser", given
+          arrays of "current" browsers from other vendors and "previous"
+          browsers (from "browser" vendor) that are older than the grace
+          period.
+
+          Let browserAPIs = <"browser" interfaces>
+          Let curAPIs = INTERSECT(<"current" interfaces>)
+          Let cutOffAPIs = <interfaces from newest releas in "previous">
+          Let oldAPIs = UNION(<"previous" releases other than "cutOff" release>)
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: interfaces from "previous"...

nit: cutOff isn't described in the docs.

File: lib/confluence/aggressive_removal.es6.js [hunk 8]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
+            this.mlang.INTERSECT(
+                currentLessOne.map(brwsr => brwsr.interfaces),
+                // Intersect with removed-before-grace-period, not-added-back:
+                // allAPIsOlderThanPostRemoval \ postRemoval \ browser.
+                this.mlang.INTERSECT(
+                    this.mlang.UNION.apply(
+                        this.mlang,
+                        prevLessOne.map(brwsr => brwsr.interfaces)),
+                    this.mlang.SET_MINUS(
+                        postRemoval.interfaces,
```
Comment (arobins) [inline lib/confluence/aggressive_removal.es6.js]: SET_MINUS is nary now, right?

File: lib/confluence/aggressive_removal.es6.js (hunk 5 above)
Comment (mdittmer) [inline lib/confluence/aggressive_removal.es6.js]: Punted to future commit. Filed #58

File: lib/confluence/aggressive_removal.es6.js (hunk 4 above)
Comment (mdittmer) [inline lib/confluence/aggressive_removal.es6.js]: I suppose this should be better documented: `compute()` is only ever invoked from super class's `run()`, which provides the latest release of each browser as the `browsers` parameter.

Should I just add a comment here? `documentation:` for `compute()` call this out more explicitly?

File: lib/confluence/aggressive_removal.es6.js (hunk 6 above)
Comment (mdittmer) [inline lib/confluence/aggressive_removal.es6.js]: I'll fix this when I do the big rename ( #58 ).

File: lib/confluence/aggressive_removal.es6.js [hunk 9]
```diff
@@ -43,97 +39,110 @@ foam.CLASS({
...
-            });
-          }));
-        }
-        return Promise.all(computeForBrowsers);
+        // Select from first current browser.
+        return firstCurrent.interfaces.select(
+            // Intersect with other current browsers to get APIs shipped by
+            // all (non-"browser" vendor) current browsers.
+            this.mlang.INTERSECT(
+                currentLessOne.map(brwsr => brwsr.interfaces),
```
Comment (mdittmer) [inline lib/confluence/aggressive_removal.es6.js]: This was to avoid a name conflict with `browser` in the outer scope.

**Pull request merged**